### ISSUE TITLE
Lint cycle: `clippy::or_fun_call` doesn't like `.or_insert(vec![])`, but `clippy::redundant_closure` doesn't like `or_insert_with(|| vec![])`

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-clippy#9522]
### Description

I was getting this warning

```
warning: use of `or_insert` followed by a call to `new`
   --> components/dada-lang/src/test_harness.rs:445:18
    |
445 |                 .or_insert(vec![])
    |                  ^^^^^^^^^^^^^^^^^ help: try this: `or_default()`
    |
    = note: `#[warn(clippy::or_fun_call)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

Per https://github.com/rust-lang/rust-clippy/issues/8574, it's a false warning, but I thought "meh, whatever, I'll do what it suggests." But then I get this...

```rust
warning: redundant closure
   --> components/dada-execute/src/heap_graph/graphviz.rs:350:33
    |
350 |                 .or_insert_with(|| vec![])
    |                                 ^^^^^^^^^ help: replace the closure with `Vec::new`: `std::vec::Vec::new`
    |
    = note: `#[warn(clippy::redundant_closure)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

This is annoying!

**Comment by kraktus** [comment]
What about `.or_insert_with(Vec::new)`?

**Comment by DzenanJupic** [comment]
Same is true for `enum` constructors
https://github.com/DzenanJupic/rustube/actions/runs/3260790886/jobs/5354665044

```
error: use of `ok_or` followed by a function call
   --> src/fetcher.rs:229:22
    |
229 |                     .ok_or(Error::Internal("The regex does not match meaningful"))?
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `ok_or_else(|| Error::Internal("The regex does not match meaningful"))`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
    = note: `-D clippy::or-fun-call` implied by `-D warnings`
```

```
error: unnecessary closure used to substitute value for `Option::None`
   --> src\fetcher.rs:227:17
    |
227 | /                 watch_html
228 | |                     .get(m.end()..)
229 | |                     .ok_or_else(|| Error::Internal("The regex does not match meaningful"))?
    | |______________________--------------------------------------------------------------------^
    |                        |
    |                        help: use `ok_or(..)` instead: `ok_or(Error::Internal("The regex does not match meaningful"))`
    |
    = note: `-D clippy::unnecessary-lazy-evaluations` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
```

**Comment by J-ZhengLi** [comment]
> Same is true for enum constructors

problem with num calls was already fixed, possibly #9610 as **[kraktus](https://github.com/kraktus)** mentioned in #9663

**Comment by J-ZhengLi** [comment]
```text
warning: redundant closure
   --> components/dada-execute/src/heap_graph/graphviz.rs:350:33
    |
350 |                 .or_insert_with(|| vec![])
    |                                 ^^^^^^^^^ help: replace the closure with `Vec::new`: `std::vec::Vec::new`
    |
    = note: `#[warn(clippy::redundant_closure)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

`or_fun_call` won't suggest this, at least not anymore. (that part got separated as `clippy::unwrap_or_default`)

fwiw, only `hm.entry(0).or_insert(vec!["abc".to_string()]).push("def".to_string());` will indeed suggest `.or_insert_with(...)` because the function call has an arg now. [play me](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=168ce1cf5f3ebfaf5e44449d0bb2c9b6), and the resulted suggestion won't triggers `redundant_closure`, so I think this could be closed as done? Feel free to re-open it if I turn out to be wrong~
